### PR TITLE
🐛 adds esType input in dashboard and fallsback on index if not presented

### PR DESCRIPTION
**Pull request by hlminh2000**

### CODE REVIEW

File: modules/server/src/projects/addType.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ import getIndexPrefix from '../utils/getIndexPrefix';
 export default async (req, res) => {
   let { es } = req.context;
   let { id } = req.params;
-  let { index, name } = req.body;
+  let { index, name, esType } = req.body;
```
Comment (jgnieuwhof) [inline modules/server/src/projects/addType.js]: maybe add the `esType` fallback after the if statement below, `if (!esType) esType = index`

File: modules/server/src/projects/addType.js [hunk 2]
```diff
@@ -69,12 +70,14 @@ export default async (req, res) => {
       let aliases = await es.cat.aliases({ format: 'json' });
       let alias = aliases?.find(x => x.alias === index)?.index;
 
+      const type = esType && esType.length ? esType : index;
```
Comment (jgnieuwhof) [inline modules/server/src/projects/addType.js]: if fallback is above no need for this

File: modules/server/src/projects/addType.js [hunk 3]
```diff
@@ -69,12 +70,14 @@ export default async (req, res) => {
...
+
       let mappings = await es.indices.getMapping({
         index: alias || index,
-        type: name,
+        type,
```
Comment (jgnieuwhof) [inline modules/server/src/projects/addType.js]: `type: esType`

File: modules/server/src/projects/addType.js [hunk 4]
```diff
@@ -69,12 +70,14 @@ export default async (req, res) => {
...
       });
 
-      let mapping = mappings[alias || index].mappings[name].properties;
+      let mapping = mappings[alias || index].mappings[type].properties;
```
Comment (jgnieuwhof) [inline modules/server/src/projects/addType.js]: `type` => `esType`

File: modules/server/src/projects/getFields.js [hunk 5]
```diff
@@ -50,7 +50,7 @@ export default async (req, res) => {
         size: 0,
         _source: false,
       });
-      const esType = projectInfo?.hits?.hits?._source?.name || index;
+      const esType = projectInfo?.hits?.hits?._source?.esType || index;
```
Comment (jgnieuwhof) [inline modules/server/src/projects/getFields.js]: wouldn't need fallback here

File: modules/server/src/startProject.js [hunk 6]
```diff
@@ -89,7 +89,7 @@ export default async function startProjectApp({
         type.name,
         {
           index: type.index,
-          es_type: type.name,
+          es_type: type.esType || type.index,
```
Comment (jgnieuwhof) [inline modules/server/src/startProject.js]: same as ^

File: modules/server/src/utils/fetchMappings.js [hunk 7]
```diff
@@ -11,15 +11,13 @@ export let fetchMapping = async ({ index, esType, es }) => {
       // TODO: return something more useful than false
       return false;
     })
-    .then(val => {
-      return { index: index, mapping: val, alias };
-    });
+    .then(val => ({ index: index, mapping: val, alias }));
```
Comment (jgnieuwhof) [inline modules/server/src/utils/fetchMappings.js]: ❤️

File: modules/server/src/utils/fetchMappings.js [hunk 8]
```diff
@@ -11,15 +11,13 @@ export let fetchMapping = async ({ index, esType, es }) => {
...
 };
 
 export let fetchMappings = ({ types, es }) => {
   return Promise.all(
-    types.map(({ index, name }) => {
-      return fetchMapping({ index, esType: name, es });
-    }),
+    types.map(({ index, name, esType }) =>
+      fetchMapping({ index, esType: esType || index, es }),
```
Comment (jgnieuwhof) [inline modules/server/src/utils/fetchMappings.js]: wouldn't need fallback here

**Comment by hlminh2000** [comment]
let's delay the deployment of this till the next time we reindex, otherwise we'd have to add `esType` into the project config manually

**Pull request merged**